django/contrib/postgres/signals.py

```diff
@@ -13,6 +13,10 @@ def register_hstore_handler(connection, **kwargs):
             register_hstore(connection.connection, globally=True, unicode=True)
         else:
             register_hstore(connection.connection, globally=True)
+        with connection.cursor() as cursor:
+            cursor.execute("""SELECT oid from pg_type WHERE typname = 'hstore'""")
+            oid = cursor.fetchone()[0]
+            connection.introspection.data_types_reverse[oid] = 'django.contrib.postgresql.fields.HStoreField'
```

Comment:
> I think that cached_property is instance-specific, not class-specific, so I'm not sure it would help.

`data_types_reverse` is only accessed on `DatabaseIntrospection` instance anyway.

> Another approach would be testing the presence of 'django.contrib.postgresql.fields.HStoreField' in data_types_reverse and only add it when missing.

One thing I like about keeping the code here instead of moving it to the backend is that it's only registered if the contrib app is installed.